### ISSUE TITLE
TransformerEncoder src_key_padding_mask does not work in eval()

### ISSUE THREAD

**Issue by AdventLee** [pytorch/pytorch#86120]
### 🐛 Describe the bug

Described here: https://discuss.pytorch.org/t/model-eval-predicts-all-the-same-or-nearly-same-outputs/155025/12

Starting with version 1.12, TransformerEncoder uses two different code paths, which perform differently between training and evaluation mode. In training, the `why_not_sparsity_fast_path` will be set, and the sparsity fast path will not be used. During eval(), if using a `src_key_padding_mask`, output will be modified:

https://github.com/pytorch/pytorch/blob/b04b2fa9aa52cacbdc9aaaf477d55b0af845ce81/torch/nn/modules/transformer.py#L271
```
            if (not why_not_sparsity_fast_path) and (src_key_padding_mask is not None):
                convert_to_nested = True
                output = torch._nested_tensor_from_mask(output, src_key_padding_mask.logical_not(), mask_check=False)
                src_key_padding_mask_for_layers = None
```

The output is usually some constant value. It seems like their might be some misalignment with the expected `src_key_padding_mask` and the `logical_not()` operation. Perhaps the mask inversion should not be performed.

### Versions

Python version: 3.8.13 (default, Mar 28 2022, 06:59:08) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 516.94
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.7.1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.3
[pip3] segmentation-models-pytorch==0.3.0
[pip3] torch==1.12.1+cu116
[pip3] torchaudio==0.12.1+cu116
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.1+cu116
[conda] efficientnet-pytorch      0.7.1                    pypi_0    pypi
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] segmentation-models-pytorch 0.3.0                    pypi_0    pypi
[conda] torch                     1.12.1+cu116             pypi_0    pypi
[conda] torchaudio                0.12.1+cu116             pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.13.1+cu116             pypi_0    pypi

**Comment by erichan1** [comment]
Shared internally. Thank you for calling this out!

**Comment by zrphercule** [comment]
Hi @AdventLee , thanks for bringing this!
For better understanding of this issue, could you please give a simple reproduce of this issue? I tried to repro it but I dont think I reproed this issue (Probably due to I misunderstood it.)
Thanks!

**Comment by AdventLee** [comment]
```
import torch
from torch import nn
import torch.optim as optim

criterion = nn.MSELoss()

pad_mask = torch.tensor([[1, 1, 0, 0]], dtype=torch.long)
layer = nn.TransformerEncoderLayer(
            d_model=2,
            dim_feedforward=4,
            nhead=2,
            batch_first=True,
            activation="gelu",
            dropout=0,
        )
encoder = nn.TransformerEncoder(layer, 2)
optimizer = optim.SGD(encoder.parameters(), lr=0.1, momentum=0.9)


encoder.train()
for i in range(100):
    encoder.train()
    optimizer.zero_grad()
    inputs = torch.cat([torch.randn(1,2,2), torch.zeros(1,2,2)], dim=1)

    outputs = encoder(inputs, src_key_padding_mask=pad_mask)

    loss = criterion(outputs[:, 0:2, :], inputs[:, 0:2, :])
    loss.backward()
    optimizer.step()

    with torch.no_grad():
        test = torch.cat([torch.randn(1,2,2), torch.zeros(1,2,2)], dim=1)
        test_train_long = encoder(test, src_key_padding_mask=pad_mask.long())
        test_train_bool = encoder(test, src_key_padding_mask=pad_mask.bool())

        encoder.eval()
        test_eval_long = encoder(test, src_key_padding_mask=pad_mask.long())
        test_eval_bool = encoder(test, src_key_padding_mask=pad_mask.bool())

        l1_train_train = nn.L1Loss()(test_train_bool[:, 0:2, :], test_train_long[:, 0:2, :]).item()
        l1_long = nn.L1Loss()(test_train_long[:, 0:2, :], test_eval_long[:, 0:2, :]).item()
        l1_bool = nn.L1Loss()(test_train_bool[:, 0:2, :], test_eval_bool[:, 0:2, :]).item()
        
        if l1_bool > 1e-4:
            print(f"Eval/Train difference in pad_mask BOOL({i:02d}): {l1_bool:.4f}")
        if l1_long > 1e-4:
            print(f"Eval/Train difference in pad_mask LONG({i:02d}): {l1_long:.4f}")
        if l1_train_train > 1e-4:
            print(f"Train/Train difference in pad_mask BOOL/LONG({i:02d}): {l1_long:.4f}")
```

Sorry it took so long, it was a bit difficult to pin down and the encoder shows the issue intermittently.

The issue seems to be related to type conversion of the pad mask. If the pad mask is boolean, then there is no issue; however, when inputting the pad mask as long, the result is different between train and eval (probably due to bool conversion at the `logical_not()`). I found you can also see differences in the results in train() as well using different types.

However, the differences are not apparent at most steps. Here is the output from a run I did:
```
Eval/Train difference in pad_mask LONG(34): 0.8490
Train/Train difference in pad_mask BOOL/LONG(34): 0.8490
Eval/Train difference in pad_mask LONG(39): 0.5622
Train/Train difference in pad_mask BOOL/LONG(39): 0.5622
Eval/Train difference in pad_mask LONG(40): 0.5577
Train/Train difference in pad_mask BOOL/LONG(40): 0.5577
Eval/Train difference in pad_mask LONG(41): 0.5929
Train/Train difference in pad_mask BOOL/LONG(41): 0.5929
Eval/Train difference in pad_mask LONG(46): 0.2728
Train/Train difference in pad_mask BOOL/LONG(46): 0.2728
Eval/Train difference in pad_mask LONG(62): 0.4571
Train/Train difference in pad_mask BOOL/LONG(62): 0.4571
Eval/Train difference in pad_mask LONG(68): 0.4367
Train/Train difference in pad_mask BOOL/LONG(68): 0.4367
Eval/Train difference in pad_mask LONG(80): 0.7433
Train/Train difference in pad_mask BOOL/LONG(80): 0.7433
```
Only a handful of steps in the loop will produce a difference. Most of the time, this code will print differences, but a few times I ran it, it did not.

**Comment by zrphercule** [comment]
@AdventLee Thank you very much, we found the reason is mainly because the old version of PyTorch does not support Long type mask. Fix and your unit test: https://github.com/pytorch/pytorch/pull/86353/files

**Comment by zrphercule** [comment]
> Hi @zrphercule Can you please explain if both mask = None and src_key_padding_mask = None in the encoder part of the transformer, will I experience this issue or not?

Hi @mariiak2021 No, this bug is actually not related to BetterTransformer; instead it is the old version of PyTorch Transformer did not handle non-bool src_key_padding_mask properly. If you runs the model in inference mode (which will lead you to BetterTransformer), or you are not using src_key_padding_mask with a non-bool dtype, you will not encounter this bug.

**Comment by malfet** [comment]
We should allow only bool dtype of `src_key_padding_mask`

**Comment by Mogady** [comment]
Is this solved? I'm getting the same issue on Colab with HuggingFace and I'm not sure what is the problem, I'm simply training an AutoModelForSequenceClassification with simple data, and whenever I do model.eval() I get exactly the same predictions for all the batches but when back to .train() it works fine
